Nuke testing class

kalisp opened the pull request:
Implemented pytest testing class for Nuke. 

Pulls prepared testing zip file from GDrive, prepares temporal DBs(avalon, openpype), opens Nuke, publishes basic workfile, compares result.

(in activated {OPENPYPE_ROOT}/.venv)
`{OPENPYPE_ROOT}/.venv/Scripts/python.exe {OPENPYPE_ROOT}/start.py runtests {OPENPYPE_ROOT}/tests/integration/hosts/nuke`

~(Contains merged PR for OP-2019- https://github.com/pypeclub/avalon-core/pull/391)~
~(Contains commit with current avalon-core)~

mkolar commented:
Task linked: [OP-2019 Create test publish class for AE](https://app.clickup.com/t/6658547/OP-2019)

mkolar commented:
Task linked: [OP-2042 nuke testing class](https://app.clickup.com/t/6658547/OP-2042)

jakubjezek001 commented on openpype/hosts/aftereffects/plugins/publish/closeAE.py:
```diff
@@ -0,0 +1,29 @@
+
+import pyblish.api
+
+from avalon import aftereffects
```
shouldn't this module be part of other PR?

jakubjezek001 commented on openpype/hosts/aftereffects/plugins/publish/extract_local_render.py:
```diff
@@ -19,10 +19,9 @@ def process(self, instance):
         staging_dir = instance.data["stagingDir"]
         self.log.info("staging_dir::{}".format(staging_dir))
```
this too is not related to Nuke, right?

jakubjezek001 commented on openpype/lib/remote_publish.py:
```diff
@@ -11,6 +11,25 @@
 from openpype.lib.plugin_tools import parse_json
 
 
+def headless_publish(log, close_plugin_name=None, is_test=False):
```
I believe this is not related either to Nuke or testing in general ;)

jakubjezek001 reviewed (changes requested):
Lot of unrelated changes in Nuke and Testing context of this PR

kalisp commented:
Cleaned up files from AE PR

kalisp commented:
Comments:
- add possibility to inject application's input arguments -- will be added to test zip similar to `env_var` folder
- add class variable to inject folder with testing data - for easier building of test class, in the future to daisy chain applications
- add implicit possibility of filling `VARIANT` - if empty look for a) latest installed version b) run tests on all installed versions

kalisp commented:
All issues after review implemented. 

Mentioned `-x` command line flag seems not to trigger `init.py`, still needs some testing.

jakubjezek001 reviewed (approved):
works

kalisp commented:
Added bunch of updates:
- new wrapper class for testing in a host - each test case in a host should extend this one
- reworked db_asserts to run all asserts, created easier method to query
- output folder is set in a better way
- added `test_data_folder`, `app_variant`, `persist` to cli
- fixed test db cleanup
- fixed published folder cleanup
- test case in Nuke finally works

kalisp commented:
All other features will be implemented in next PRs.
